### PR TITLE
feat: provider models for signing transactions

### PR DESCRIPTION
**Pull request by PhearZero**
# Overview

- adds ARC27 Provider messages
- fixes tsconfig build issues
- exports Provider as subpackage `@algorandfoundation/algo-models/provider`

# Issues

- https://algorandfoundation.atlassian.net/browse/ENG-523

### CODE REVIEW

File: examples/esm/index.js [hunk 1]
```diff
@@ -1,7 +1,9 @@
 
 import { AlgorandTransactionCrafter } from '@algorandfoundation/algo-models'
+import {SignTransactionsParams} from "@algorandfoundation/algo-models/provider";
 
 let atc = new AlgorandTransactionCrafter('testnet-v1.0', 'SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=')
 let pay = atc.pay(123, 'R2HYOMCQIIVZLYYSVQ7RRDME5QB2NNACNIYEYIIN3NCTEKHEVE4RJRNWN4', 'R2HYOMCQIIVZLYYSVQ7RRDME5QB2NNACNIYEYIIN3NCTEKHEVE4RJRNWN4')
 //console.log(Buffer.from(pay.get().encode()).toString('base64'))
 console.log(pay)
+console.log(SignTransactionsParams)
```
Comment (ehanoc) [inline examples/esm/index.js:9]: Is this file meant to be a test case?

File: lib/provider/api.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+const chars =
+    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';
+export const INVALID_BASE64URL_INPUT = 'Invalid base64url input';
+
+
+/**
+ * Bytes to Base64URL
+ * @param {Uint8Array| ArrayBuffer} arr Bytes to convert to URL safe Base64
+ */
+export function toBase64URL(arr: Uint8Array | ArrayBuffer): string {
```
Comment (ehanoc) [inline lib/provider/api.ts:20]: Isn't there something that already does this?

**Review by ehanoc — CHANGES_REQUESTED** [review]
Looking good! Just a couple comments

File: examples/esm/index.js (hunk 1 above)
Comment (PhearZero) [reply examples/esm/index.js:9]: Exactly, it's for manual QA to ensure the three artifacts are produced properly.

- cjs uses /dist/cjs/*
- esm uses /dist/esm/*
- types/typescript uses /dist/types/*

File: lib/provider/api.ts (hunk 2 above)
Comment (PhearZero) [reply lib/provider/api.ts:20]: Sadly there is not a cross platform native way for Base64 URL. `Buffer` only exists in Node.js and not the Browser.

Uint8Array/ArrayBuffer are available in both contexts but lack any string serializers like Buffer. The existing packages I could find where out of date and small enough in scope to just implement our own version.

Comment (ehanoc) [reply lib/provider/api.ts:20]: cool makes sense

File: examples/esm/index.js (hunk 1 above)
Comment (ehanoc) [reply examples/esm/index.js:9]: Can't they be through a regular test suite? Someone importing index{.js, .ts, ) would run those bits by default, no?

Comment (PhearZero) [reply examples/esm/index.js:9]: Absolutely is possible, although It's non trivial to distribution test with Node.js/Browser/Typescript/package.json combinations. We are using it to make sure it works in a third party consumer context and why they have their own `package.json`

The examples folder is less than ideal but it's a good place to manually install the package in different contexts until we have a test suite for the distributions.

**Review by ehanoc — APPROVED** [review]

**Pull request merged**